Trakt/ScheduledTasks/SyncLibraryTask.cs

```diff
@@ -385,140 +465,218 @@ public async Task ExecuteAsync(IProgress<double> progress, CancellationToken can
+                        currentProgress += percentPerRequest;
+                        progress.Report(currentProgress);
+
+                        LogTraktResponseDataContract(dataContracts, TraktItemType.episode);
                     }
+
+                    dataContracts = (await _traktApi.SendLibraryUpdateAsync(
+                        episodes,
+                        traktUser,
+                        collected ? EventType.Add : EventType.Remove,
+                        cancellationToken)
+                    .ConfigureAwait(false))
+                    .ToList();
+
+                    currentProgress += percentPerRequest;
+                    progress.Report(currentProgress);
+
+                    LogTraktResponseDataContract(dataContracts, TraktItemType.episode);
+                }
+                catch (ArgumentNullException argNullEx)
+                {
+                    _logger.LogError(argNullEx, "ArgumentNullException handled sending episodes to trakt.tv");
                 }
                 catch (Exception e)
                 {
-                    _logger.LogError(e, "Error updating episode play states");
+                    _logger.LogError(e, "Exception handled sending episodes to trakt.tv");
                 }
-
-                progress.Report(100);
             }
         }
 
-        private async Task SendEpisodeCollectionUpdates(
-            bool collected,
+        private async Task SendEpisodePlaystateUpdates(
+            bool seen,
             TraktUser traktUser,
-            ICollection<Episode> collectedEpisodes,
-            ISplittableProgress<double> progress,
+            List<Episode> episodes,
+            IProgress<double> progress,
+            double currentProgress,
+            double availablePercent,
             CancellationToken cancellationToken)
         {
-            _logger.LogInformation("Episodes to add to Collection: {Count}", collectedEpisodes.Count);
-            if (collectedEpisodes.Count > 0)
+            if (episodes.Count > 0)
             {
+                _logger.LogInformation("Episodes to set {State}watched: {Count}", seen ? string.Empty : "un", episodes.Count);
                 try
                 {
-                    var dataContracts =
-                        await _traktApi.SendLibraryUpdateAsync(
-                                collectedEpisodes,
-                                traktUser,
-                                collected ? EventType.Add : EventType.Remove,
-                                cancellationToken)
-                            .ConfigureAwait(false);
-                    if (dataContracts != null)
+                    var dataContracts = new List<TraktSyncResponse>();
+                    var percentPerRequest = availablePercent / (episodes.Count / 100.0);
+
+                    // Force update trakt.tv if we have more than 100 movies in the queue due to API
+                    while (episodes.Count > 100)
```

Comment:
This reduces the amount of lists created

```diff
Index: Trakt/ScheduledTasks/SyncLibraryTask.cs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/Trakt/ScheduledTasks/SyncLibraryTask.cs b/Trakt/ScheduledTasks/SyncLibraryTask.cs
--- a/Trakt/ScheduledTasks/SyncLibraryTask.cs	(revision 73dc33dd88616b6f74bff4d99cce22a166dce405)
+++ b/Trakt/ScheduledTasks/SyncLibraryTask.cs	(date 1650586098241)
@@ -588,9 +588,10 @@
                     var percentPerRequest = availablePercent / (episodes.Count / 100.0);
 
                     // Force update trakt.tv if we have more than 100 movies in the queue due to API
-                    while (episodes.Count > 100)
+                    var offset = 0;
+                    while (offset < episodes.Count)
                     {
-                        var episodesToSend = episodes.Take(100).ToList();
+                        var episodesToSend = episodes.GetRange(offset, 100);
                         dataContracts = await _traktApi.SendEpisodePlaystateUpdates(
                             episodesToSend,
                             traktUser,
@@ -598,7 +599,7 @@
                             cancellationToken)
                         .ConfigureAwait(false);
 
-                        episodes.RemoveRange(0, 100);
+                        offset += 100;
                         currentProgress += percentPerRequest;
                         progress.Report(currentProgress);
 

```

Reply:
I don't think that's true. It will create a new list regardless, but GetRange is more efficient.